### ISSUE TITLE
clang -fPIC creates wrong relocation type for wasm32

### ISSUE THREAD

**Issue by TerrorJack** [llvm/llvm-project#63354]
Minimal repro:

```c
void internal_memalign();
void *dlmemalign() { return internal_memalign; }
```

Compile and link with:

```shell
clang --target=wasm32 -fPIC -c test.c -o test.o
wasm-ld --shared --no-gc-sections test.o -o test.wasm
```

Linking step fails with:

```console
wasm-ld: warning: creating shared libraries, with -shared, is not yet stable
wasm-ld: error: test.o: relocation R_WASM_TABLE_INDEX_SLEB cannot be used against symbol `internal_memalign`; recompile with -fPIC
```

This can be reproduced by latest LLVM build from current `main` branch.

**Comment by TerrorJack** [comment]
Ah ha, I found the culprit: https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/WebAssembly/WebAssemblyTargetMachine.cpp#L106

If I comment away this line I can get wasi-sdk to emit proper PIC objects now. And I do think this logic should get away; I understand the current dylink stack is only used by emscripten yet, but it shouldn't get in the way of other users that attempt to implement their own wasm dynamic linker. What do you all think?

**Comment by TerrorJack** [comment]
Closed via 55e199a2c9f4b218499733d60129deffa0a025fe.
